proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,754 @@
+be backwards-compatible. Notably, before this feature, the kind of a type family
+only shows up when using ``StandaloneKindSignatures`` or in GHCi when using the
+``:kind`` command.
+
+We propose a new flag, ``-fprint-explicit-matchabilities``, similar to
+``-fprint-explicit-runtime-reps``, that only shows the matchability information
+to users who ask.
+
+Alternatives
+------------
+
+There are a number of alternative decisions regarding the specific
+details of the proposal.
+
+1.  Instead of matchability polymorphism,
+    a subsumption relationship could be considered between the two arrows.
+    This approach has been fully formalised by Richard Eisenberg in his
+    `thesis <http://www.cis.upenn.edu/~sweirich/papers/eisenberg-thesis.pdf>`_,
+    The main drawback of that approach is that inference would suffer compared
+    to the scheme outlined above. Matchability polymorphism also fits more
+    cleanly into the existing constraint solver mechanism. Alternatively, we
+    could stick to the simple defaulting scheme mentioned in the section about
+    annotation-free arrows. But...
+
+2.  Type inference with the "simple" matchability defaulting scheme is
+    incomplete. Take following program ::
+      nested :: a b ~ c Id => b Bool
+      nested = False
+
+    Initially, the matchabilities of ``a``, ``b`` and ``c`` are all
+    instantiated with unification variables, and there are no further
+    steps. So they are all defaulted to be matchable, at which point
+    the equality can be decomposed, and we learn that
+    ``(b :: Type -> @M Type) ~ (Id :: Type -> @U Type)``. This way, ``nested`` cannot be
+    called, because no such ``b`` exists.
+
+    Instead, we could do something more clever by defaulting
+    matchabilities in dependency order, but it's not obvious if this
+    additional complexity would be worth it.
+
+3.  We could make different choices for the syntax, regarding how we annotate
+    arrows with matchabilities or the particular names around the
+    ``Matchability`` type.
+
+4.  Data constructors could be considered to have matchable types. This would
+    make promotion more unified, as promoted constructors have matchable kinds.
+    This is quite appealing, but doing so would require additional engineering
+    effort. Either we would need to introduce matchability-polymorphic
+    term-level functions, or, perhaps more realistically, eta-expand all data
+    constructor applications to demote them to unmatchable.
+
+5.  When a kind signature is given, we make the choice of not generalising the
+    matchabilities, which differs from the treatment of kind variables. Consider
+    the following program ::
+
+      type A :: Proxy a -> Type
+      type family A
+
+    The inferred kind is ``A :: forall {k} (a :: k). Proxy a -> @M Type``, so the
+    kind of the type variable ``a`` did get generalised, but the matchability of
+    the arrow didn't (note that ``A`` is nullary). An alternative option would be
```

Comment:
Nitpick: `A` isn't technically nullary, as it is indexed by the kind variable `a`.